app/services/irs_attempts_api/tracker.rb

```diff
@@ -1,20 +1,31 @@
 module IrsAttemptsApi
   class Tracker
-    attr_reader :session_id, :enabled_for_session
+    attr_reader :session_id, :enabled_for_session, :request, :user, :sp
 
-    def initialize(session_id:, enabled_for_session:)
-      @session_id = session_id
+    def initialize(session_id:, request:, user:, sp:, enabled_for_session:)
+      @session_id = session_id # IRS session ID
+      @request = request
+      @user = user
+      @sp = sp
       @enabled_for_session = enabled_for_session
     end
 
     def track_event(event_type, metadata = {})
       return unless enabled?
 
+      event_metadata = {
+        user_agent: request&.headers['User-Agent'],
+        unique_session_id: user&.unique_session_id,
+        user_uuid: ServiceProviderIdentity.where(service_provider: sp.issuer, user_id: user.id).take&.uuid,
```

Comment:
Confusingly, AgencyIdentityLinker doesn't appear to support looking things up the way I expected. I had expected `sp_identity_from_uuid_and_sp` would take a User uuid, but it does not.

Doing this right here is a little gross, though.

Reply:
This is the wrong table... this UUID is used as a legacy fallback, we want `AgencyIdentity#uuid` (because the Agency UUID is preferred, so two apps at the same agency can share a UUID space)

But also yes, there *should* be better methods for accessing, I think it would be OK if we added `AgencyIdentity.for(user:, service_provider:).uuid` method in this PR

Reply:
Oh, I was thrown off by this method:

```ruby
  def self.sp_identity_from_uuid_and_sp(uuid, service_provider)
    ai = AgencyIdentity.where(uuid: uuid).take
    criteria = if ai
                 { user_id: ai.user_id, service_provider: service_provider }
               else
                 { uuid: uuid, service_provider: service_provider }
               end
    ServiceProviderIdentity.where(criteria).take
  end
```

It left me thinking that this was all just a wrapper on ServiceProviderIdentity. I like this; I'll add this.

Reply:
@zachmargolis Actually, I'm still a little puzzled.

As I read it, AgencyIdentityLinker suggests that not all user have an AgencyIdentity yet, but it creates them on the fly from a ServiceProviderIdentity if there isn't one. Is that logic still necessary? I.e., should `AgencyIdentity.for(user:, service_provider:)` be a simple lookup, or does it need to fall back to ServiceProviderIdentity if no record is found?

Reply:
The most correct behavior is to fall back to `ServiceProviderIdentity` (the core issue is that some service providers, especially in lower environments, don't have an Agency so that's why we call back to the plain ServiceProviderIdentity UUID)

Here's how we use the one for the OpenID connect `sub` claim:

https://github.com/18F/identity-idp/blob/2a23ace2101802f98320ca219117eead1038d810/app/presenters/openid_connect_user_info_presenter.rb#L13

https://github.com/18F/identity-idp/blob/2a23ace2101802f98320ca219117eead1038d810/app/presenters/openid_connect_user_info_presenter.rb#L33-L35

So I don't love using the `AgencyIdentityLinker` because it could create a link when one might not be there, but reusing one of those lines is probably the easiest way to maintain parity